dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.cpp

```diff
@@ -39,32 +39,63 @@ SegmentReadTaskScheduler::~SegmentReadTaskScheduler()
 }
 
 void SegmentReadTaskScheduler::add(const SegmentReadTaskPoolPtr & pool)
+{
+    std::lock_guard lock(add_mtx);
+    submitPendingPool(pool);
+}
+
+void SegmentReadTaskScheduler::addPool(const SegmentReadTaskPoolPtr & pool)
 {
     assert(pool != nullptr);
-    Stopwatch sw_add;
-    // `add_lock` is only used in this function to make all threads calling `add` to execute serially.
-    std::lock_guard add_lock(add_mtx);
-    add_waittings.fetch_add(1, std::memory_order_relaxed);
-    // `lock` is used to protect data.
-    std::lock_guard lock(mtx);
-    add_waittings.fetch_sub(1, std::memory_order_relaxed);
-    Stopwatch sw_do_add;
     read_pools.emplace(pool->pool_id, pool);
 
     const auto & tasks = pool->getTasks();
     for (const auto & [seg_id, task] : tasks)
     {
         merging_segments[seg_id].push_back(pool->pool_id);
     }
+}
+
+void SegmentReadTaskScheduler::addPools(const SegmentReadTaskPools & pools)
+{
+    for (const auto & pool : pools)
+    {
+        addPool(pool);
+    }
+    LOG_INFO(log, "Added, pool_ids={}, pool_count={}", pools, read_pools.size());
+}
+
+void SegmentReadTaskScheduler::submitPendingPool(SegmentReadTaskPoolPtr pool)
+{
+    assert(pool != nullptr);
+    if (pool->getPendingSegmentCount() <= 0)
+    {
+        LOG_INFO(pool->getLogger(), "Ignored for no segment to read, pool_id={}", pool->pool_id);
+        return;
+    }
+    Stopwatch sw;
+    std::lock_guard lock(pending_mtx);
+    pending_pools.push_back(pool);
     LOG_INFO(
         pool->getLogger(),
-        "Added, pool_id={} block_slots={} segment_count={} pool_count={} cost={:.3f}us do_add_cost={:.3f}us", //
+        "Submitted, pool_id={} segment_count={} pending_pools={} cost={}ns",
         pool->pool_id,
-        pool->getFreeBlockSlots(),
-        tasks.size(),
-        read_pools.size(),
-        sw_add.elapsed() / 1000.0,
-        sw_do_add.elapsed() / 1000.0);
+        pool->getPendingSegmentCount(),
+        pending_pools.size(),
+        sw.elapsed());
+}
+
+void SegmentReadTaskScheduler::reapPendingPools()
+{
+    SegmentReadTaskPools pools;
+    {
+        std::lock_guard lock(pending_mtx);
+        pools.swap(pending_pools);
+    }
+    if (!pools.empty())
+    {
+        addPools(pools);
```

Comment:
We can directly put the for-loop here and remove the function `addPools`
```suggestion
        for (const auto & pool : pools)
        {
            addPool(pool);
        }
        LOG_INFO(log, "Added, pool_ids={}, pool_count={}", pools, read_pools.size());
```